salt/modules/yumpkg.py

```diff
@@ -1513,18 +1513,20 @@ def install(name=None,
             else:
                 pkgstr = pkgpath
 
-            # Lambda to trim the epoch from the currently-installed version if
-            # no epoch is specified in the specified version
-            norm_epoch = lambda x, y: x.split(':', 1)[-1] \
-                if ':' not in y \
-                else x
             cver = old_as_list.get(pkgname, [])
             if reinstall and cver:
                 for ver in cver:
-                    ver = norm_epoch(ver, version_num)
-                    if salt.utils.versions.compare(ver1=version_num,
+                    if diff_attr:
+                        # Since diff_attr was provided, the version info
+                        # is a dict.
+                        ver2 = ver['version']
+                    else:
+                        # No diff_attr was provided, so version is directly
+                        # in ver.
+                        ver2 = ver
```

Comment:
I would write it easier and safer:
```python
ver2 = diff_attr and ver.get('version') or ver
```

Reply:
I for one find the original version clearer - especially because your version, not being parenthesised, forces the reader to think about operator precedence.

I do agree, though, that using an expression is safer. How about:

```python
ver2 = ver['version'] if diff_attr else ver
```

Reply:
+1 for @moio

Reply:
@moio so it is:
```python
ver2 = ver['version'] if diff_attr else ver
```
I am fine with this.

Reply:
As soon as there is a `diff_attr`, the version is added to the dict. I saw that in the code and just verified it.

Reply:
@brejoc OK, in `pkg_resource::format_pkg_list`, if `attr`, then `version` is pushed, yes. In this case type of the `ver` variable changes between tuple and dict. Then I would do it differently:
```python
ver2 = isinstance(ver, dict) and ver.get('version') or ver
```

Reply:
Because I consider overwriting the variable used in the loop as a bad practice. It's a constant stream of failure. The loop is short, but it's a habit of mine that worked well for me.

Reply:
The variable is only used once. I see no need to allocate memory twice. Really.

Reply:
@brejoc I actually see it as a bugmaking: you supposed to extract the version, which is a tuple of things. It is either put in a structure or made directly. I agree that this design itself is very strange, but we have what we have. Now you created "var2" like this is another version. I don't like that and would prefer to get rid of what one doesn't need.